modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTests.java

```diff
@@ -157,11 +153,79 @@ public void testParseTemplateAsSingleStringWithConditionalClause() throws IOExce
         TemplateScript TemplateScript = compiled.newInstance(script.getParams());
         assertThat(TemplateScript.execute(), equalTo("{ \"match_all\":{} }"));
     }
 
+    private static class TestReflection {
+
+        private final int privateField = 1;
+
+        public final int publicField = 2;
+
+        private int getPrivateMethod() {
+            return 3;
+        }
+
+        public int getPublicMethod() {
+            return 4;
+        }
+
+        @Override
+        public String toString() {
+            return List.of(privateField, publicField, getPrivateMethod(), getPublicMethod()).toString();
+        }
+    }
+
+    /**
+     * BWC test for some odd reflection edge-cases. It's not really expected that customer code would be exercising this,
+     * but maybe it's out there! Who knows!?
+     *
+     * If we change this, we should *know* that we're changing it.
+     */
+    @SuppressWarnings({ "deprecation", "removal" })
+    public void testReflection() {
+        Map<String, Object> vars = Map.of("obj", new TestReflection());
+
+        {
+            // non-reflective access calls toString
+            String templateString = "{{obj}}";
+            String o = qe.compile(null, templateString, TemplateScript.CONTEXT, Map.of()).newInstance(vars).execute();
+            assertThat(o, equalTo("[1, 2, 3, 4]"));
+        }
+        {
+            // accessing a field/method that *doesn't* exist will give an empty result
+            String templateString = "{{obj.missing}}";
+            String o = qe.compile(null, templateString, TemplateScript.CONTEXT, Map.of()).newInstance(vars).execute();
+            assertThat(o, equalTo(""));
+        }
+        {
+            // accessing a private field that does exist will give an empty result
+            String templateString = "{{obj.privateField}}";
+            String o = qe.compile(null, templateString, TemplateScript.CONTEXT, Map.of()).newInstance(vars).execute();
+            assertThat(o, equalTo(""));
+        }
+        {
+            // accessing a private method that does exist will give an empty result
+            String templateString = "{{obj.privateMethod}}";
+            String o = qe.compile(null, templateString, TemplateScript.CONTEXT, Map.of()).newInstance(vars).execute();
+            assertThat(o, equalTo(""));
+        }
+        {
+            // accessing a public field that does exist will give an empty result
+            String templateString = "{{obj.publicField}}";
+            String o = qe.compile(null, templateString, TemplateScript.CONTEXT, Map.of()).newInstance(vars).execute();
+            assertThat(o, equalTo(""));
+        }
+        {
+            // accessing a public method that does exist will give an empty result
+            String templateString = "{{obj.publicMethod}}";
+            String o = qe.compile(null, templateString, TemplateScript.CONTEXT, Map.of()).newInstance(vars).execute();
+            assertThat(o, equalTo(""));
```

Comment:
Are the last two cases (public field & method) the current behaviour?

Reply:
Ah wait, the reflection invocation failed (not allowed to use `setAccessible(...)`) so accessing fields or methods never worked here.

Maybe add a bwc yaml test to show case this behaviour also?

Reply:
@martijnvg precisely!

> Maybe add a bwc yaml test to show case this behaviour also?

Yup, will do.

Reply:
Added via d9ba896